### ISSUE TITLE
[Bug]: why the logits is different between 0.4.1 and 0.4.2

### ISSUE THREAD

**Issue by sitabulaixizawaluduo** [vllm-project/vllm#4740]
### Your current environment

```text
The output of `python collect_env.py`
```

### 🐛 Describe the bug

```
from vllm import LLM, SamplingParams
sampling_params = SamplingParams(temperature=0, max_tokens=2048)
llm = LLM(model="Llama-3-8B",tensor_parallel_size=4, trust_remote_code=True)
outputs = llm.generate(prompts=prompts, sampling_params=sampling_params)
for output in outputs:
    prompt = output.prompt
    generated_text = output.outputs[0].text
    print(prompt + generated_text)         
```                                              
I use vllm 0.4.1 and 0.4.2 run this code, and I print logits in sampler.py, but I found the logits is different between 0.4.1 and 0.4.2 when the  tensor_parallel_size is 4, but is the same when tensor_parallel_size is 2 or 1. 
**TP4 first token logits:**
0.4.2
![image](https://github.com/vllm-project/vllm/assets/71002153/313484a2-3a60-4b91-8243-ada4b49a864f)
0.4.1 
![image](https://github.com/vllm-project/vllm/assets/71002153/20906e58-5568-4cc1-9dc1-4dbd15789c8b)
**TP2 first token logits:**
0.4.2
![image](https://github.com/vllm-project/vllm/assets/71002153/871fb6ce-238b-4752-8adb-f64267302e46)
0.4.1
![image](https://github.com/vllm-project/vllm/assets/71002153/6c874e36-c90e-47f5-a7af-2f0ec4e6e78e)

**Comment by sitabulaixizawaluduo** [comment]
I useFlashAttention backend with flash-attn==2.5.2
